examples/model_selection/plot_grid_search_refit_callable.py

```diff
@@ -79,49 +101,293 @@ def best_low_complexity(cv_results):
 
 grid = GridSearchCV(
     pipe,
     cv=10,
-    n_jobs=1,
+    n_jobs=1,  # increase this on your machine to use more physical cores
     param_grid=param_grid,
     scoring="accuracy",
     refit=best_low_complexity,
+    return_train_score=True,
 )
+
+# %%
+# Load the digits dataset and fit the model
+# ---------------------------------------
+
 X, y = load_digits(return_X_y=True)
 grid.fit(X, y)
 
+# %%
+# Visualize the results
+# -------------------
+#
+# We'll create a bar chart showing the test scores for different numbers of PCA
+# components, along with horizontal lines indicating the best score and the
+# one-standard-deviation threshold.
+
 n_components = grid.cv_results_["param_reduce_dim__n_components"]
 test_scores = grid.cv_results_["mean_test_score"]
 
-plt.figure()
-plt.bar(n_components, test_scores, width=1.3, color="b")
+# Create a polars DataFrame for better data manipulation and visualization
+results_df = pl.DataFrame(
+    {
+        "n_components": n_components,
+        "mean_test_score": test_scores,
+        "std_test_score": grid.cv_results_["std_test_score"],
+        "mean_train_score": grid.cv_results_["mean_train_score"],
+        "std_train_score": grid.cv_results_["std_train_score"],
+        "mean_fit_time": grid.cv_results_["mean_fit_time"],
+        "rank_test_score": grid.cv_results_["rank_test_score"],
+    }
+)
+
+# Sort by number of components
+results_df = results_df.sort("n_components")
 
+# Calculate the lower bound threshold
 lower = lower_bound(grid.cv_results_)
-plt.axhline(np.max(test_scores), linestyle="--", color="y", label="Best score")
-plt.axhline(lower, linestyle="--", color=".5", label="Best score - 1 std")
-
-plt.title("Balance model complexity and cross-validated score")
-plt.xlabel("Number of PCA components used")
-plt.ylabel("Digit classification accuracy")
-plt.xticks(n_components.tolist())
-plt.ylim((0, 1.0))
-plt.legend(loc="upper left")
 
+# Get the best model information
 best_index_ = grid.best_index_
+best_components = n_components[best_index_]
+best_score = grid.cv_results_["mean_test_score"][best_index_]
+
+# Add a column to mark the selected model
+results_df = results_df.with_columns(
+    pl.when(pl.col("n_components") == best_components)
+    .then(pl.lit("Selected"))
+    .otherwise(pl.lit("Regular"))
+    .alias("model_type")
+)
+
+# Get the number of CV splits from the results
+n_splits = sum(
+    1
+    for key in grid.cv_results_.keys()
+    if key.startswith("split") and key.endswith("test_score")
+)
+
+# Extract individual scores for each split
+test_scores = np.array(
+    [
+        [grid.cv_results_[f"split{i}_test_score"][j] for i in range(n_splits)]
+        for j in range(len(n_components))
+    ]
+)
+train_scores = np.array(
+    [
+        [grid.cv_results_[f"split{i}_train_score"][j] for i in range(n_splits)]
+        for j in range(len(n_components))
+    ]
+)
+
+# Calculate mean and std of test scores
+mean_test_scores = np.mean(test_scores, axis=1)
```

Comment:
We could use `np.nanmean`, `np.nanstd` and `np.nanmax` to make this code robust to single class folds for which ROC AUC would not be defined.

This would only happen with non-stratified CV and imbalanced data (which is not the case with Iris) but as explained earlier, better make this code easy to reuse on imbalanced data.

Reply:
Closing this as ROC AUC is not a valid choice for multiclass problems and accuracy is always defined.